src/System.Security.Cryptography.Algorithms/tests/HKDFTests.cs

```diff
@@ -0,0 +1,611 @@
+        }
+
+        [Fact]
+        public void Rfc5869DeriveKeySpanNonSensicalHash()
+        {
+            byte[] ikm = new byte[20];
+            byte[] okm = new byte[17];
+            Assert.Throws<ArgumentOutOfRangeException>(() => HKDF.DeriveKey(new HashAlgorithmName("foo"), ikm, okm, Array.Empty<byte>(), Array.Empty<byte>()));
+        }
+
+        [Theory]
+        [MemberData(nameof(GetRfc5869TestCases))]
+        public void Rfc5869DeriveKeySpanTamperIkmTests(Rfc5869TestCase test)
+        {
+            test.Ikm[0] ^= 1;
+            byte[] okm = new byte[test.Okm.Length];
+            HKDF.DeriveKey(test.Hash, test.Ikm, okm, test.Salt, test.Info);
+            Assert.NotEqual(test.Okm, okm);
+        }
+
+        [Fact]
+        public void Rfc5869DeriveKeySpanOkmMaxSizePlusOne()
+        {
+            byte[] ikm = new byte[20];
+            byte[] okm = new byte[20 * 255 + 1];
+            Assert.Throws<ArgumentException>(() => HKDF.DeriveKey(HashAlgorithmName.SHA1, ikm, okm, Array.Empty<byte>(), Array.Empty<byte>()));
+        }
+
+        [Theory]
+        [MemberData(nameof(GetRfc5869TestCasesWithNonEmptySalt))]
+        public void Rfc5869DeriveKeySpanTamperSaltTests(Rfc5869TestCase test)
+        {
+            test.Salt[0] ^= 1;
+            byte[] okm = new byte[test.Okm.Length];
+            HKDF.DeriveKey(test.Hash, test.Ikm, okm, test.Salt, test.Info);
+            Assert.NotEqual(test.Okm, okm);
+        }
+
+        [Theory]
+        [MemberData(nameof(GetRfc5869TestCasesWithNonEmptyInfo))]
+        public void Rfc5869DeriveKeySpanTamperInfoTests(Rfc5869TestCase test)
+        {
+            test.Info[0] ^= 1;
+            byte[] okm = new byte[test.Okm.Length];
+            HKDF.DeriveKey(test.Hash, test.Ikm, okm, test.Salt, test.Info);
+            Assert.NotEqual(test.Okm, okm);
+        }
+
+        public static IEnumerable<object[]> GetRfc5869TestCases()
+        {
+            foreach (Rfc5869TestCase test in Rfc5869TestCases)
+            {
+                yield return new object[] { test };
+            }
+        }
+
+        public static IEnumerable<object[]> GetRfc5869TestCasesWithNonEmptySalt()
+        {
+            foreach (Rfc5869TestCase test in Rfc5869TestCases)
+            {
+                if (test.Salt != null && test.Salt.Length != 0)
+                {
+                    yield return new object[] { test };
+                }
+            }
+        }
+
+        public static IEnumerable<object[]> GetRfc5869TestCasesWithNonEmptyInfo()
+        {
+            foreach (Rfc5869TestCase test in Rfc5869TestCases)
+            {
+                if (test.Info != null && test.Info.Length != 0)
+                {
+                    yield return new object[] { test };
+                }
+            }
+        }
+
+        private static Rfc5869TestCase[] Rfc5869TestCases => new Rfc5869TestCase[7]
```

Comment:
Unless any test cases are altering data, you only need to make this once.

```suggestion
        private static Rfc5869TestCase[] Rfc5869TestCases { get; } = new[]
```

If the tests are altering data, you could make them first make a copy before altering.

Reply:
I initially did it this way but found it much easier to write tests when you can just alter the data

Reply:
Fair enough.  You could still use `new[]` instead of `new Rfc5869TestCase[7]` and let the compiler simply count the number of commas to make the array the proper size.

Reply:
I ended up changing it. I left explicit number since the rfc test cases are number and this double checks I haven't missed any